### DISCUSSION TITLE
Add support for OSC control surfaces like TouchOSC

### DISCUSSION THREAD

**Discussion by Cheesiq** [File-New-Project/EarTrumpet#696] [Ideas]
Would be awesome to see support for OSC since it's becoming the new standard for control surfaces.

MIDI support would also be nice, however it limits you to having 16 channels whereas OSC is limitless.

**Comment by riverar** [comment]
Hey @CheesePwn, is your thinking here that a hardware product would implement OSC communication and talk to EarTrumpet to control app volumes?

**Comment by Cheesiq** [comment]
> Hey @CheesePwn, is your thinking here that a hardware product would implement OSC communication and talk to EarTrumpet to control app volumes?

This is correct yes.

**Comment by riverar** [comment]
@CheesePwn Cool, thanks. Will read up on this!

**Comment by flibble666** [comment]
If NK2Tray supported more device it would save alot of work.
https://github.com/ho0ber/NK2Tray

Also what about adding rtpMIDI this would allow apps like TouchDAW.
https://www.tobias-erichsen.de/software/rtpmidi.html
http://www.humatic.de/htools/touchdaw/

**Comment by Cheesiq** [comment]
@flibble666 Actual Solution's Power Mixer has a MIDI option, however it is paid software and I'm just thinking of a feature that would be beneficial to all users but also free. There's no option other than NK2Tray, which is limited in devices, as you pointed out.

**This will be a long comment about the various solutions for controlling application volumes via a MIDI app/device or OSC supported app/device.**

**Power Mixer** (Paid) - Easiest solution
Power Mixer is an awesome tool, the only drawback being the price. That's why I suggested this feature in the first place so the barrier to entry isn't price and would be beneficial to EarTrumpet users.
http://www.actualsolution.com/power-mixer/

**NK2Tray**
NK2Tray is good, and if they can investigate more devices, that'd be great! I do recall asking them about OSC support on their Discord but they were saying about the codebase not being the best to implement features on at this moment in time.

**Protokol** (Free)
For monitoring MIDI and OSC protocols:
https://hexler.net/products/protokol

**TouchOSC** (Paid)
Modular touch control surface for OSC & MIDI
https://hexler.net/products/touchosc

MIDI is still overly complex in setup though if you don't have experience in sound design. Luckily, the solutions I've used are slightly fiddly, but works in practice.

I did try looking into rtpMIDI but it confused me on the setup so I gave up on that idea. I just use loopMIDI for creating virtual MIDI ports and MIDI-OX for routing/looping.

**Major differences between MIDI and OSC**
- OSC is able to communicate through IP, and apps such as TouchOSC.
- MIDI is limited to 16 channels (usually), and you would have to loop the MIDI signal to be able to use it on multiple programs.

**Sidenotes**
An app called [TouchPortal](https://www.touch-portal.com/) is planning to add OSC/MIDI support in the future, but only solves issue for tablets and phones, not hardware MIDI devices. The project is a different scope to EarTrumpet anyway, but I felt it should be mentioned as they are doing cool stuff over there!

**Conclusion**
I still think it would be beneficial for EarTrumpet to have these features baked in, but ymmv.

Cheers for the insight! 👍

**Comment by flibble666** [comment]
Thank you  i will check out Power Mixer  thx for the info. or wait ti see if   TouchPortal  adds midi support.
